docs/proposals/gpu-support.md

```diff
@@ -0,0 +1,263 @@
+scheduled, requesting one device per pod. The scheduler puts them all on
+different machines, spreading the load, perhaps by design. Then a new job comes
+in, requiring two devices per pod, but it can't schedule anywhere, because all
+we can find, at most, is one unused device per node.
+
+For v0, we'll use [a scheduler extender](../design/scheduler_extender.md). An
+extender runs outside of `kube-scheduler` and is consulted before picking a
+candidate node to run on. It can keep track of GPU resource usage on its own,
+either in a local store or by managing its own set of labels on nodes. The
+obvious benefits of flexibility and increased experimentation are countered by
+the drawback of adding more work on the administrator's plate. They have one
+more component to set up, maintain and monitor. Because the extender enters the
+picture only late into the game, it does not obviate the need for a rescheduler.
+
+#### Scheduler extender design
+
+The extender can be called at two stages during scheduling:
+
+- filtering, to select viable nodes
+- ranking, to assign viable nodes a priority
+
+For v0 we'll use the former only. The extender is passed information about the
+pod being scheduled and a list of nodes. Presumably, the user will restrict pods
+to GPU-enabled instances only: because extenders are called after predicates
+such as matching node labels, the candidates will be a subset of the whole
+cluster. Filtering will act on the node list only if the pod has requested GPUs
+(TBD: use annotations or [fully-qualified resource
+types](../design/resources.md#resource-types)?].
+
+The extender will return just one from all the nodes that it knows to have
+available GPU resources. It won't return multiple nodes, because the API does
+not inform the extender about the node that is eventually picked. Without the
+node's identity, resources can't be reserved permanently. It is possible to
+obtain the information indirectly by placing a watch on the pod, but there might
+be race conditions between delivery of the notification and the next incoming
+extender request. Resources are thus committed right away to the pod on that
+specific node (this can be a temporary lease, in case the pod can't schedule
+there for other reasons). Because of that, the GPU extender should be the last
+in the scheduler policy configuration. Actually, a watch will still be created,
+but, instead, it will be used to return the resources used by the pod when it
+terminates.
+
+Nodes with available resources are handed out in a circular fashion in order to
+
+- approximate round robin scheduling
+- increase the chances of forward progress when the same pod is passed to the
+  extender multiple times (e.g. because of other constraints not being met)
+
+### The runtime
+
+Once we know where to run the container, it's time to set up its environment. We
+rely on the [nvidia-docker plugin](https://github.com/NVIDIA/nvidia-docker),
```

Comment:
Done, but there's still a lot more in the nvidia-docker section than in the general one.

I'd love to have more generic infrastructure from the beginning, but in reality I can't find anyone trying to use non-NVIDIA devices in containers. The closest is this, which hasn't been touched in a year and only deals with creating OpenCL images:

https://github.com/chesterkuo/docker-gpgpu

I doubt it's a coincidence that both AWS and Azure support NVIDIA only. Same with Mesos. I can probably find a contact at Intel and inquire. Not sure about AMD. It looks like only NVIDIA has libraries (in Go!) and plugins to abstract (a little) the hardware and the drivers/libraries controlling it, to expose all that to a container engine.